src/include/firebird/FirebirdInterface.idl

```diff
@@ -1557,16 +1558,39 @@ interface Int128 : Versioned
 
 interface ReplicatedRecord : Versioned
 {
+	// Return count of fields
+	uint getCount();
+	// Return name of field. NULL is returned on error.
+	const string getField(uint index);
+	// Return dtype_* constants or dtype_unknown if field not exists.
+	uint getType(uint index);
```

Comment:
Why storage internals should be exposed to replication plugin if message types can be correspondent to them?

AFAIK only place dtypes are exposed are for UDFs and they are deprecated. I do not see a sense in reinsert them as public thing.

Reply:
Nope, they are exposed to public in dsc_pub.h which is included into ibase.h and Interface.h.

Reply:
Exposed for what usage currently?

Reply:
Frankly, I have no idea. It is too old thing.

Reply:
So you confirm my point it should not continue to be exposed as there is something that already replaces it externally.

Reply:
I considered that, but found it inappropriate. Record format has to be repacked into a message format for every record modification and this may be a noticeable performance issue. It would be great to have messages more flexible (so that they could be accessed from different storage formats using the same API), but currently it's not really so. Perhaps field values could be accessed using getOffset() returning an internal record position (do we have the same alignment rules for record fields and message elements?), but getNullOffset() breaks everything -- representing NULLs as SHORT is absolutely dumb requirement that's inherited from the legacy code. Calls like getRelation() / getOwner() / getAlias() are also redundant here, but that's not a big deal - empty string can be returned.

Reply:
That said, I have no objections for getType() returning the SQL type rather than dtype (for consistency with other API).

Reply:
>  I have no objections for getType() returning the SQL type rather than dtype

So do I.

Another question: should getLength() also follow this rule? The length of dtype_varying is two bytes bigger that length of SQL_VARYING for the same field.

Reply:
I think yes, it should. If we return SQL_VARYING as a type, then its size should also correspond to SQL_VARYING.

Reply:
> That said, I have no objections for getType() returning the SQL type rather than dtype (for consistency with other API).

That is what I was suggesting mainly.